toolkit/tools/srpmpacker/srpmpacker.go

```diff
@@ -969,16 +971,29 @@ func hydrateFromRemoteSource(fileHydrationState map[string]bool, newSourceDir st
 
 		url := network.JoinURL(srcConfig.sourceURL, fileName)
 
-		// With 10 attempts, initial delay of 1 second, and a backoff factor of 2.0, this will
-		// take at most ~5 minutes before giving up.
+		// Limit the number of concurrent network operations by pushing a struct{} into the channel. This will block until
+		// another operation completes and removes the struct{} from the channel.
+		if netOpsSemaphore != nil {
+			netOpsSemaphore <- struct{}{}
```

Comment:
You should pass the cancellation signal into this function and then use a `select` statement here so that the function can exit early.

I suspect this will allow you to get rid of the `drainChannel` stuff.

Reply:
You might also want to strongly consider using the [context](https://pkg.go.dev/context) package and its associated coding patterns at some point. (But probably as a separate change.)

Reply:
Huh... that context stuff does look useful.
I wasn't able to get the cancel signal to work well in two spots, but I'll give it another go (since agreed, its a bit cleaner).